content/blog/rag-pipelines-from-scratch/index.md

```diff
@@ -0,0 +1,423 @@
+
+![RAG Paper diagram](rag-paper-image.png)
+
+As you can see in the image above (taken directly from the original paper), a system such as RAG is made of two parts: one that finds text snippets that are relevant to the question asked by the user and a generative model, usually an LLM, that rephrases the snippets into a coherent answer for the question.
+
+Let's build one of these with Haystack 2.0!
+
+> 💡 *Do you want to see this code in action? Check out the Colab notebook [here](https://colab.research.google.com/drive/1vX_2WIRuqsXmoPMsJbqE45SYn21yuDjf?usp=drive_link) or the [gist](https://gist.github.com/ZanSara/cad6f772d3a894058db34f566e2c4042).
+
+
+> ⚠️ **Warning:** *This code was tested on `haystack-ai==0.88.0`. Haystack 2.0 is still unstable, so later versions might introduce breaking changes without notice until Haystack 2.0 is officially released. The concepts and components however stay the same.*
+
+# Generators: Haystack's LLM components
+
+As every NLP framework that deserves its name, Haystack supports LLMs in different ways. The easiest way to query an LLM in Haystack 2.0 is through a Generator component: depending on which LLM and how you intend to query it (chat, text completion, etc...), you should pick the appropriate class.
+
+We're going to use ChatGPT for these examples, so the component we need is [`GPTGenerator`](https://github.com/deepset-ai/haystack/blob/main/haystack/preview/components/generators/openai/gpt.py). Here is all the code required to use it to query OpenAI's ChatGPT:
+
+```python
+from haystack.preview.components.generators.openai.gpt import GPTGenerator
+
+generator = GPTGenerator(api_key=api_key)
+generator.run(prompt="What's the official language of France?")
+# returns {"replies": ['The official language of France is French.']}
+```
+You can select your favorite OpenAI model by specifying a `model_name` at initialization, for example, `gpt-4`. It also supports setting an `api_base_url` for private deployments, a `streaming_callback` if you want to see the output generated live in the terminal, and optional `kwargs` to let you pass whatever other parameter the model understands, such as the number of answers (`n`), the temperature (`temperature`), etc.
+
+Note that in this case, we're passing the API key to the component's constructor. This is unnecessary: `GPTGenerator` can read the value from the `OPENAI_API_KEY` environment variable and also from the `api_key` module variable of [`openai`'s SDK](https://github.com/openai/openai-python#usage).
+
+Right now, Haystack supports HuggingFace models through the [`HuggingFaceLocalGenerator`](https://github.com/deepset-ai/haystack/blob/f76fc04ed05df7b941c658ba85adbf1f87723153/haystack/preview/components/generators/hugging_face/hugging_face_local.py#L65) component, and many more LLMs are coming soon.
```

Comment:
Please check if these links are still valid. Also, we can now add `HuggingFaceTGIGenerator` here

Reply:
Consider that the Colab is pinned to a older Haystack 2.0 version, it needs review if we want to use the latest. I'd rather publish as is and make a review round when we publish the beta, to bring everything up-to-date. Otherwise we need to update this every day :see_no_evil: